tests/testthat/test-shinytest2-teal-widgets.R

```diff
@@ -0,0 +1,218 @@
+    table_buttons <-
+      app$get_html_rvest(table_buttons_selector) %>%
+      rvest::html_elements("button")
+    testthat::expect_length(table_buttons, 2)
+    # Check is the first one is a toggle button.
+    testthat::expect_equal(
+      table_buttons[[1]] %>%
+        html_attr("data-toggle"),
+      "dropdown"
+    )
+    # First button has specific font-awesome icon.
+    dwnl_button <- app$active_module_element("downbutton-dwnl")
+    testthat::expect_equal(
+      app$get_html_rvest(dwnl_button) %>%
+        rvest::html_element("i") %>%
+        rvest::html_attr("class"),
+      "fas fa-download"
+    )
+
+    # Click the first TABLE button.
+    app$click(selector = dwnl_button)
+    app$wait_for_idle(timeout = default_idle_timeout)
+
+    # Review the content of the toggle.
+    testthat::expect_equal(
+      app$active_module_element_text("downbutton-file_format-label"),
+      "File type"
+    )
+
+    file_format_text <- app$active_module_element_text("downbutton-file_format > div")
+    testthat::expect_match(file_format_text, "formatted txt\n", fixed = TRUE)
+    testthat::expect_match(file_format_text, "csv\n", fixed = TRUE)
+    testthat::expect_match(file_format_text, "pdf\n", fixed = TRUE)
+
+    testthat::expect_equal(
+      app$active_module_element_text("downbutton-file_name-label"),
+      "File name (without extension)"
+    )
+
+    # QUESTION - NEED HELP
+    # How to get
+    # value="table_20240321_142134"
+    # out of
+    # <input id="teal-main_ui-root-module_with_table_with_settings-module-downbutton-file_name" type="text" class="shiny-input-text form-control shinyjs-resettable shiny-bound-input" value="table_20240321_142134" data-shinyjs-resettable-id="teal-main_ui-root-module_with_table_with_settings-module-downbutton-file_name" data-shinyjs-resettable-type="Text" data-shinyjs-resettable-value="table_20240321_142134">
+    # for
+    # app$get_html_rvest(app$active_module_element("downbutton-file_name"))
+
+    # TODO - NEED HELP
+    # do not use div:nth-child(3)
+    pagination <- "downbutton-dwnl > li > div > div:nth-child(3) > div.paginate-ui  > div.form-group.shiny-input-container"
+    pagination_class <- gsub("#", "#dropdown-menu-", app$active_module_element(pagination))
+    pagination_text <- app$get_text(pagination_class)
+    testthat::expect_match(pagination_text, "Paginate table:\n", fixed = TRUE)
+    testthat::expect_match(pagination_text, "lines / page\n", fixed = TRUE)
+
+    download_button <- app$get_html_rvest(app$active_module_element("downbutton-data_download > i"))
+    testthat::expect_equal(
+      download_button %>%
+        html_node("i") %>%
+        html_attr("class"),
+      "fas fa-download"
+    )
+    testthat::expect_equal(
+      download_button %>%
+        html_node("i") %>%
+        html_attr("aria-label"),
+      "download icon"
+    )
+
+    # TODO - NEED HELP
+    # click second radio button
+    # app$click(selector = "input[value~=csv]")
```

Comment:
To check that is being hidden needs javascript :-\

```r
    app$click(selector = "input[value='.csv']")
    !any(
      unlist(
        app$get_js(
          sprintf(
            "Array.from(document.querySelectorAll('%s')).map(el => el.checkVisibility())", 
            pagination_class
          )
        )
      )
    )
```

Reply:
This basically performs a `lapply` on the selector results and use `checkVisibility()` function ([docs](https://developer.mozilla.org/en-US/docs/Web/API/Element/checkVisibility))

`NodeList` object in javascript doesn't support the `map` method, so it needs to convert to Array first... _(don't get me started on js)_

Reply:
I think I couldn't get `app$click(selector = "input[value='.csv']")` to actually make a click

Reply:
Alrighty, but `app$get_text(pagination_class)` still returns content of the pagination panel, even though it's not visible anymore. So now I understand your comment about checking the Visibility of the element. Totally

Reply:
bece513b0db346ae66ca13c6b20ffb365d10fc0a